code/controllers/subsystem/persistent_paintings.dm

```diff
@@ -140,52 +172,84 @@ SUBSYSTEM_DEF(persistent_paintings)
+		current_data =  migrate_to_version_2(current_data)
+
+	return current_data
+
+/datum/controller/subsystem/persistent_paintings/proc/migrate_to_version_1(list/current_data)
+	fcopy("data/paintings.json","data/paintings_migration_backup_0.json") //Better safe than losing all metadata
+	var/list/result = list()
+	result["version"] = 1
+	var/list/data = list()
+	// Squash categories into tags
+	for(var/category in current_data)
+		for(var/old_data in current_data[category])
+			var/duplicate_found = FALSE
+			for(var/list/entry in data)
+				if(entry["md5"] == old_data["md5"])
+					entry["tags"] |= category
+					duplicate_found = TRUE
+					break
+			if(duplicate_found)
+				continue
+			var/old_png_path = "data/paintings/[category]/[old_data["md5"]].png"
+			var/new_png_path = "data/paintings/images/[old_data["md5"]].png"
+			fcopy(old_png_path,new_png_path)
+			fdel(old_png_path)
+			var/icon/painting_icon = new(new_png_path)
+			var/width = painting_icon.Width()
+			var/height = painting_icon.Height()
+			var/list/new_data = list()
+			new_data["md5"] = old_data["md5"]
+			new_data["title"] = old_data["title"] || "Untitled Artwork"
+			new_data["creator_ckey"] = old_data["ckey"] || ""
+			new_data["creator_name"] = "Anonymous"
+			new_data["creation_date"] = time2text(world.realtime) // Could use creation/modified file helpers in rustg
+			new_data["creation_round_id"] = GLOB.round_id
+			new_data["tags"] = list(category,"Migrated from version 0")
+			new_data["patron_ckey"] = ""
+			new_data["patron_name"] = ""
+			new_data["credit_value"] = 0
+			new_data["width"] = width
+			new_data["height"] = height
+			new_data["medium"] = "Spraypaint on canvas" //Let's go with most common tool.
+			data += list(new_data)
+	result["paintings"] = data
+	//We're going to save this immidiately this is non-recoverable operation
+	var/json_file = file("data/paintings.json")
+	fdel(json_file)
+	WRITE_FILE(json_file, json_encode(result))
+	return result
+
+/datum/controller/subsystem/persistent_paintings/proc/migrate_to_version_2(list/current_data)
+	fcopy("data/paintings.json","data/paintings_migration_backup_1.json") //Better safe than sorry
+	current_data["version"] = 2
+	for(var/painting_data in current_data["paintings"])
+		painting_data["tags"] |= "Migrated from version 1"
+		var/credit_value = painting_data["credit_value"]
+		var/list/possible_frame_types = get_available_frames(credit_value, only_current_tier = TRUE)
+		painting_data["frame_type"] = pick(possible_frame_types) || "simple"
+	return current_data
+
+/**
+ * returns a list of cosmetic frames which patronage tier values are are within credit_value.
+ * If only_current_tier is TRUE, the list will be populated only by the ones from the highest tier.
+ */
+/datum/controller/subsystem/persistent_paintings/proc/get_available_frames(credit_value, only_current_tier = FALSE)
+	var/list/frame_candidates = list()
+	var/current_tier_patronage = 0
```

Comment:
Is this used for anything, you can just continue the loop here ?

Reply:
This var is pretty much only used if only_current_tier is TRUE. When only_current_tier is TRUE, only the frame(s) whose required required value is closest to that of the `credit_value` argument (of course, without surpassing it) will be in the `frame_candidates` list, which  is the return value of the proc.
Here how it works: We have Alice going to Bob's soup store. She has 10 coins and she wants to buy ONE of the most expensive soup she can afford. The shop has different kind of soups. Some cost less, some cost more. In order to make her choice, Alice has to find which soup/s is/are the most expensive. For the sake of the argument, let's say she is going to check one soup at a time and Bob won't help. In this case, she would have to keep track of the price of the most expensive soup she has checked so far, on top of the list of soups she would buy, so that she can compare it with the next one and make a choiche whether ignore it, clear the list of soups she would buy and/or add it to such list...
This is pretty what the `current_tier_patronage` variable is far (but for frames): Keeping track of the value of the most expensive frame(s) that we can afford out of those we have checked so far.